google-cloud-spanner/src/main/java/com/google/cloud/spanner/connection/SingleUseTransaction.java

```diff
@@ -298,7 +299,8 @@ private ApiFuture<ResultSet> executeDmlReturningAsync(
                 writeTransaction.run(
                     transaction ->
                         DirectExecuteResultSet.ofResultSet(
-                            transaction.executeQuery(update.getStatement(), options)));
+                            transaction.executeQuery(
+                                update.getStatement(), appendLastStatement(options))));
```

Comment:
Quick clarification: Since a SingleUseTransaction can only execute one query, do we implicitly mark that statement as the last statement? The changes in this file seem to be related to this, correct?

What happens if we don’t mark it as the last statement in SingleUseTransaction?
Since the customer cannot run additional queries anyway, I’m trying to understand the benefit of marking it as the last statement.

Reply:
Note: This is only relevant in read/write transactions, so in this case only for DML statements with a THEN RETURN clause, which are also executed using `executeQuery(..)`.

Your observation that we automatically mark a statement in a SingleUseTransaction as the last_statement is correct. We do this, because we know that this will be the only statement in this transaction. By giving Spanner this hint, Spanner can internally optimise the execution of this statement by skipping some verifications, that are also executed during the Commit call.

And more in general: By supplying this hint to Spanner, Spanner could in the future also consider other possible optimizations that are possible when it knows that the current statement will be the last in a read/write transaction.